fdbcli/VersionEpochCommand.actor.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * VersionEpochCommand.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "boost/lexical_cast.hpp"
+
+#include "fdbcli/fdbcli.actor.h"
+
+#include "fdbclient/IClientApi.h"
+
+#include "flow/Arena.h"
+#include "flow/FastRef.h"
+#include "flow/ThreadHelper.actor.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+namespace fdb_cli {
+
+const KeyRef versionEpochSpecialKey = LiteralStringRef("\xff\xff/management/version_epoch");
+
+ACTOR Future<bool> versionEpochCommandActor(Reference<IDatabase> db, std::vector<StringRef> tokens) {
+	if (tokens.size() == 1 || tokens.size() == 3) {
+		state Reference<ITransaction> tr = db->createTransaction();
+		state int64_t versionEpoch = -1;
+		loop {
+			try {
+				Optional<Value> versionEpochVal = wait(safeThreadFutureToFuture(tr->get(versionEpochSpecialKey)));
+				ASSERT(versionEpochVal.present()); // should always return a default value
+				versionEpoch = boost::lexical_cast<int64_t>(versionEpochVal.get().toString());
+				break;
+			} catch (Error& e) {
+				wait(safeThreadFutureToFuture(tr->onError(e)));
+			}
+		}
+
+		if (tokens.size() == 1) {
+			printf("Current version epoch: %" PRId64 "\n", versionEpoch);
+			return true;
+		} else if (tokens.size() == 3) {
+			int64_t v;
+			int n = 0;
+			if (sscanf(tokens[2].toString().c_str(), "%" SCNd64 "%n", &v, &n) != 1 || n != tokens[2].size()) {
+				printUsage(tokens[0]);
+				return false;
+			}
+
+			state int64_t newVersionEpoch = versionEpoch;
+			if (tokencmp(tokens[1], "set")) {
+				newVersionEpoch = v;
+			} else if (tokencmp(tokens[1], "add")) {
```

Comment:
I wonder if this is potentially error-prone; adding to the epoch essentially has the effect of decreasing your versions. I could imagine somebody wanting to add 100 seconds to their version and calling `add 100` instead of `add -100`. I wonder if we could name this differently such that it was hard to get wrong.

Reply:
I think it could be fine to increase/decrease the version if the command suggested that, but I probably wouldn't do that with a name like `versionepoch add`. To do something like that, you would basically want something like `adjust_target_version +/-100`. This is probably how the user thinks about it for relative adjustments, and the fact that there is a reference epoch being modified under the hood is sort of an implementation detail.

For setting the absolute version, though, I think we unfortunately need to express this as the reference epoch. This is because we may want to set this on multiple clusters, and we need some way to do so that isn't a function of the current time. I suppose we could also expose a way to set the target version directly, but we would need to have both options.

Reply:
With the existing `advanceversion` command, this quickly becomes confusing if there is another new `adjust_target_version` command. I see what you are saying, and we could do something like this:

* `versionepoch` prints the current version epoch (this is the current behavior)
* `versionepoch setepoch <epoch>`, just to make it super explicit that the epoch is being set and not the version. It's a little redundant and `versionepoch set <epoch>` could work as well
* `versionepoch adjustepoch <adjustment>` same as above, could just use `adjust` as well
* `adjust_target_version <offset>` which changes the target version but under the hood is actually modifying the version epoch

Another alternative is to try and keep everything related to modifying the version under a single top level command. We already have a `getversion` command though, so adding this additional command could cause confusion.

* `version setepoch <absolute_epoch>`
* `version adjustepoch <adjustment>`
* `version getepoch`
* `version adjusttarget <adjustment>`

I think I am more inclined to keep the `versionepoch` command, but maybe expose a new command to clearly adjust the version backwards/forwards without thinking about the epoch, as you mentioned. Thoughts?

Reply:
Another usability aspect is that the version epoch is currently a value measured in seconds (matching how the Unix time is typically measured as seconds since epoch). But versions match to microseconds, so issuing a command like `advance the version epoch by 100` means advance the version epoch by 100, thus decreasing the expected version by 100,000,000. But saying `advance the version by 100` would only advance the version by, well, 100. So it might also make more sense to measure the version epoch in microseconds.

Reply:
I think we could probably be ok not having a command to do relative adjustments to the epoch, only to the target version. It may also be ok to express adjustments in terms of seconds, though perhaps that could be confusing unless the command made it explicit.